spec/struct.dd

```diff
@@ -372,108 +378,122 @@ $(H2 $(LEGACY_LNAME2 Struct-Constructor, struct-constructor, Struct Constructors
+        ---
+        struct S
+        {
+            int j = 1;
+            long k = 2;
+            this(long k)
+            {
+                this.k = k;
+            }
+            this(int i)
+            {
+                // At this point: j=1, k=2
+                $(CODE_HIGHLIGHT this)(6); // delegating constructor call
+                // At this point: j=1, k=6
+                j = i;
+                // At this point: j=i, k=6
+            }
+        }
+        ---
+        )
 
-void main(string[] args)
-{
-    auto b = B(10);
-    // b.a = 10; does not compile here, A does not define opAssign(int).
-}
-------
+        $(P The following restrictions apply to struct construction:)
 
-$(P The program above prints the line $(CODE "[= A.this(10) =]"). Anywhere else
-attempting to assign an integer to an object of type `A` would count as an
-assignment (and is not compilable because `A` does not define `opAssign(int)`).)
+        $(OL
+        $(LI If a constructor's code contains a delegate constructor call, all
+        possible execution paths through the constructor must make exactly one
+        delegate constructor call:
 
-$(P Finding the first assignment in lexical order is flow-sensitive upon the
-`if` statement. Consider a change to struct `B` in the previous example as
-follows:)
+        $(SPEC_RUNNABLE_EXAMPLE_FAIL
+        ---
+        struct S
+        {
+            int a;
+            this(int i) { }
 
-------
-struct B
-{
-    A a;
-    this(int x)
-    {
-        if (x < 0)
-            a = -x;
-        else
-            a = x;
-    }
-}
-------
+            this(char c)
+            {
+                c || this(1); // error, not on all paths
+            }
 
-$(P This code issues a constructor call on each branch of the `if` statement.
-However, such flow sensitivity is limited. There is no static or dynamic
-analysis of coverage of the `if` statement. For example:)
+            this(wchar w)
+            {
+                (w) ? this(1) : this('c'); } // ok
 
-------
-struct B
-{
-    A a;
-    this(int x)
-    {
-        if (false) a = 0; // constructor call even if never covered
-        a = x; // error, cannot assign
-    }
-}
-------
+            this(byte b)
+            {
+                foreach (i; 0 .. b)
+                {
+                    this(1);  // error, inside loop
+                }
+            }
+        }
+        ---
+        )
+        )
 
-$(P Also, member assignments inside loops are never considered constructors,
-even if it can be determined statically that the loop executes at most once.
-Example:)
+        $(LI It is illegal to refer to $(D this) implicitly or explicitly
+        prior to making a delegate constructor call.)
 
-------
-struct B
-{
-    A a;
-    this(int x)
-    {
-        foreach (i; 0 .. x ? 0 : 1) a = i; // error, cannot assign
-    }
-}
-------
+        $(LI Once the delegate constructor returns, all fields are considered
+        constructed.)
```

Comment:
I would point out the fact that the compiler does not enforce that the delegate constructor *actually* initializes the field.

Reply:
What I mean is that this is valid code:
```d
struct A
{   
    int a, b, c;
    this(int a)
    {
        this.a = a;
    }
 
    this(int a, int b)
    {
        this(a);
    }
}
```